### PR TITLE
Run CI (Build Image and committer PRs) on self-hosted runner

### PR DESCRIPTION
**Pull request by ashb**
The queue for GitHub actions is starting to get painful.

This PR changes it to run any build of main (push or scheduled), or a PR
from a committer on the self-hosted runner.

However since this setting is configured _in_ the repo, it could be
changed by a PR, so we can't rely on that; we also use a custom build of
the self-hosted runner binary that enforces PRs are from contributors,
otherwise it will fail the build

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: .github/workflows/ci.yml
Comment (potiuk) [inline .github/workflows/ci.yml:72]: I think this can be replaced by a pre-requisite step (before build-info) that could produce a string output that will be used as "fromJson()" in the "runs-on". I can make a fixup with it.

Comment (ashb) [reply .github/workflows/ci.yml:72]: I thought about that but there is one major downside: that would have to run on "public" queues, meaning it would still be at the mercy of long queue time

Comment (potiuk) [reply .github/workflows/ci.yml:72]: Not really. See https://github.com/astronomer/airflow/pull/1197

Comment (ashb) [reply .github/workflows/ci.yml:72]: Ohhh I see, yeah cool.

**Comment by ashb** [comment]
Well my security code is running, but it's perhaps a bit over-zealous 

>  Running job on this worker disallowed by security policy

File: .github/workflows/ci.yml
Comment (ashb) [reply .github/workflows/ci.yml:72]: Done in https://github.com/apache/airflow/pull/13730/commits/939a7fe948f2d917d1c1eca5100a1161cd5ab1b3

I kept it in the "Build Info" job, as each job needs a new worker, so having a worker tied up to run a very simple shell script seemed wasteful, especially as this falls reasonably cleanly under the "work out build info" description of the existing step

Comment (potiuk) [reply .github/workflows/ci.yml:72]: Yep. Easier indeed.

**Comment by ashb** [comment]
Buuuger.

>   Pulling rabbitmq (rabbitmq:3.7)...
>   toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: > https://www.docker.com/increase-rate-limit

**Comment by ashb** [comment]
Hmmm, there are a lot of images pulled on that node I wouldn't expect:

```
busybox                                                       1.26.2                c30178c5239f        3 years ago         1.11MB
busybox                                                       1.26.2-uclibc         c30178c5239f        3 years ago         1.11MB
busybox                                                       1.26.1                8a4698b1752b        4 years ago         1.11MB
busybox                                                       1.26.1-uclibc         8a4698b1752b        4 years ago         1.11MB
busybox                                                       1.26.1-musl           91f9a04c4751        4 years ago         1.23MB
busybox                                                       1.26.1-glibc          415ca0076b84        4 years ago         4.2MB
busybox                                                       1.26.0                1efc1d465fd6        4 years ago         1.1MB
busybox                                                       1.26.0-uclibc         1efc1d465fd6        4 years ago         1.1MB
busybox                                                       1.26.0-musl           d87346805225        4 years ago         1.22MB
busybox                                                       1.26.0-glibc          60d524cdf996        4 years ago         4.19MB
busybox                                                       1.25                  e02e811dd08f        4 years ago         1.09MB
busybox                                                       1.25-uclibc           e02e811dd08f        4 years ago         1.09MB
busybox                                                       1.25.1                e02e811dd08f        4 years ago         1.09MB
busybox                                                       1.25.1-uclibc         e02e811dd08f        4 years ago         1.09MB
busybox                                                       1.25-musl             733eb3059dce        4 years ago         1.21MB
busybox                                                       1.25.1-musl           733eb3059dce        4 years ago         1.21MB
busybox                                                       1.25-glibc            21c16b6787c6        4 years ago         4.19MB
busybox                                                       1.25.1-glibc          21c16b6787c6        4 years ago         4.19MB
busybox                                                       1.25.0                2b8fd9751c4c        4 years ago         1.09MB
busybox                                                       1.25.0-uclibc         2b8fd9751c4c        4 years ago         1.09MB
busybox                                                       1.25.0-musl           4f4228f93f67        4 years ago         1.21MB
busybox                                                       1.25.0-glibc          c76ece91fb87        4 years ago         4.19MB
busybox                                                       1.24                  47bcc53f74dc        4 years ago         1.11MB
busybox                                                       1.24-uclibc           47bcc53f74dc        4 years ago         1.11MB
busybox                                                       1.24.2                47bcc53f74dc        4 years ago         1.11MB
busybox                                                       1.24.2-uclibc         47bcc53f74dc        4 years ago         1.11MB
busybox                                                       1.24-musl             f8bba50164e1        4 years ago         1.2MB
busybox                                                       1.24.2-musl           f8bba50164e1        4 years ago         1.2MB
busybox                                                       1.24-glibc            54df49495ae4        4 years ago         4.18MB
busybox                                                       1.24.2-glibc          54df49495ae4        4 years ago         4.18MB
busybox                                                       1.24.1                8b17efa7e876        4 years ago         1.11MB
busybox                                                       1.24.1-uclibc         8b17efa7e876        4 years ago         1.11MB
busybox                                                       1.24.1-musl           cfef8d8972f1        4 years ago         1.2MB
busybox                                                       1.24.1-glibc          bd9cd091f6e9        4 years ago         4.18MB
busybox                                                       1-ubuntu              d34ea343a882        5 years ago         4.35MB
busybox                                                       1.21-ubuntu           d34ea343a882        5 years ago         4.35MB
busybox                                                       1.21.0-ubuntu         d34ea343a882        5 years ago         4.35MB
busybox                                                       1.24.0                e4a4ff8a080d        5 years ago         1.11MB
busybox                                                       1.23                  a84c36ecc374        5 years ago         1.1MB
busybox                                                       1.23.2                a84c36ecc374        5 years ago         1.1MB
```

Any idea why all of those are being pulled down @potiuk ?

**Comment by ashb** [comment]
I have made each node pre-load the following images from an S3 bucket:

```
mongo               3                            430ca129e871        4 days ago          446MB
python              3.6-slim-buster              eb06c4651c8a        6 days ago          111MB
python              3.7-slim-buster              36b080f01d18        6 days ago          112MB
python              3.8-slim-buster              be5d294735c6        6 days ago          113MB
python              3.9-slim-buster              edd87973afe0        6 days ago          114MB
postgres            9.6                          2a89dff7e432        6 days ago          200MB
postgres            13                           acf5fb8bfd76        6 days ago          314MB
mysql               5.7                          cc8775c0fe94        6 days ago          449MB
mysql               8                            d4c3cafb11d5        6 days ago          545MB
cassandra           3.0                          b768847ae761        7 weeks ago         388MB
apache/airflow      krb5-kdc-server-2020.10.08   0f7d9aeebddc        3 months ago        495MB
apache/airflow      openldap-2020.07.10-2.4.50   3f41a12550c4        6 months ago        262MB
redis               5.0.1                        c188f257942c        2 years ago         94.9MB
rabbitmq            3.6                          8325cf2abbb1        2 years ago         125MB
```

🤞🏻

**Comment by potiuk** [comment]
Yes. Docker-composes of our still use a number of "official" images for integration and we seem to hit the "November" limits of DockerHub (we were protected from those so far apparently because GitHub public runners are exempted from thee limits).

You. can see all of them here https://github.com/apache/airflow/tree/master/scripts/ci/docker-compose

We already brought some of the images under "apache" organisation:
* image: apache/airflow:krb5-kdc-server-2020.10.08
* image: apache/airflow:openldap-2020.07.10-2.4.50
* image: apache/airflow:presto-2020.10.08

Those "apache" images are exempted from the rate limit introduced by DockerHub in November (Infra confirmed that in one of the discussions we had. The limits are described here: https://docs.docker.com/docker-hub/download-rate-limit/  and ASF has apparently a separate agreement with DockerHub about it).

There are still quite a few images in our build that are just pulled from the "official" sources. They are used for "integration test". 

* image: mysql:${MYSQL_VERSION}
* image: postgres:${POSTGRES_VERSION}
* image: cassandra:3.0
* image: mongo:3
* image: apachepinot/pinot:latest
* image: rabbitmq:3.7
* image: redis:5.0.1

Each of those images might have different base images (busybox/alpine etc.) so it's quite expected to have a number of bae images pulled. As in Github Actions, all the images are run on a "fresh" machine, those images are pulled every time integration tests are run. And we have also a cleeanup stage that prunes the docker engine for every step, to save the space on the Github Runners (we used to have disk space problems, on those runners that we mitigated by 'system prune" for docker engine. Right now - basically ever time we run tests, those images are pulled (and this is anyhow needed for GitHub Runners because they are always "fresh") 

The limits (for unauthenticated pulls) are rather severe and we seem to be hitting those when we use self-hosted runners. Those limits are per "public IP address" not per  user, and we might be sharing the IPs with others (or we can ourselves hit the  limits. The limit is 200 container image requests per 6 hours (https://www.docker.com/pricing) so it is rather easy to hit it .

Those limits are not applicable for pulling images from GitHub public runners (they exempted their IP addresses from that), that's why we have not experienced it yet.

There are few mitigations possible:

1) skipping the "free space" step on self-hosted runners. That will make pulls to the DockerHub far less frequent, but this also means that the docker engine on our instances might grow uncontrollably and we need to do periodic pruning or restarting the instances (depending on where/how docker engine is running).  Possibly changing "docker system prune"  and removing "--all" there might help, but then we need to likely do some extra regular cleanup for those instances (GitHub Runners have the benefit that they are being maintained/started /reset by GitHub so we did not have to worry about it).

2)  moving all the images we pull to "apache/airflow" DockerHub - similarly as we did with  openldap/presto/kerberos. Those images are exempt from he DockerHub limits and we can pull them as many times we want. This seems like the easiest solution. Rather than building those images, wee can simply pull->push them. We just need to figure out sensible naming convention because right now we can only store "apache/airflow" images, but we already have precendents with `apache/airflow:openldap-2020.07.10-2.4.50` and the like.

3)  authenticating the requests from CI to DockerHub (we need a secret setup by INFRA for that).

4)  moving all the images to our "GitHub Packages"  - the legacy container registry - there GITHUB_TOKEN might be used to authenticate (we already do that to pull our images that we share with all the jobs).

5)  moving all the images we pull to "gchr.io" - the new registry enabled last week by INFRA has no limits (it is free for public images https://github.blog/2020-09-01-introducing-github-container-registry/) and it allows for public visibility of the images. Here are migrations steps: https://docs.github.com/en/packages/guides/migrating-to-github-container-registry-for-docker-images. We might simply pull/push the right images we are using and change the docker-compose images to point to it. The drawback is that currently the access management is manual (using the GitHub UI) -  but we have limited number of images so that might as well work.

**Comment by ashb** [comment]
> 1. skipping the "free space" step on self-hosted runners

is what I'm trying right now.

The plain is to scale down EC2 instances if they aren't doing anything, and after a bit of time, so (if we can get it working, remains to be seen) that is probably a workable solution.

**Comment by ashb** [comment]
> Each of those images might have different base images (busybox/alpine etc.) so it's quite expected to have a number of bae images pulled

It might have layers, but that won't result in pulling those as base images, so something else was causing those images to be explicitly pulled.

**Comment by potiuk** [comment]
> > Each of those images might have different base images (busybox/alpine etc.) so it's quite expected to have a number of bae images pulled
> 
> It might have layers, but that won't result in pulling those as base images, so something else was causing those images to be explicitly pulled.

This is one of the reasons we did 'docker system prune --all' in the first place because Public Github Runners had a number of images pre-pulled always when we entered it. Likely it's the Github Runner makes sure they are pulled before starting a job.

**Comment by ashb** [comment]
I haven't been able to test the "Build Image" step as until this PR is merged that will only run with master code. So if approved I will merge tomorrow morning when I have time to fix any fall out.

**Comment by potiuk** [comment]
@ashb  -> just to reiterate - those are very same failures I used to observe in many CI environments earlier. The biggest benefit of having full control over the CI machines is the ability to switch to in-memory-only operations.

Switching to `tmpfs` solves the need of solving those kind of I/O failures in the cloud environment as well ups speed up everything several times (sometimes even 10 times).

I spent hourse  days on solving these kind of problems in the past and they are gone entirely with several times overall build time improvement when we switch to tmpfs for both the source code location (for any kind of node setup/precompiling/parsing/execution) and docker files (for actual image building/container execution etc.). 

I could have saved days  of work last time i've done similar workarouds for spurious I/O failures - and then I stopped worrying about those kind of problems altogether when switched to tmpfs. This is like being in another world entirely. 

Just saying.

**Comment by ashb** [comment]
Jarek: sound good - any pointers on how to use tmpfs for this sort of work load?

**Comment by potiuk** [comment]
> Jarek: sound good - any pointers on how to use tmpfs for this sort of work load?
> 
> Most things I find quickly when talking about tmpfs and docker are using tmpfs for a single volume, but we want more here.
> 
> (Also if we're using tmpfs we can probably not need a local SSD anymore either)

Yeah. The thing is that docker will also nicely work when the whole 'docker root is a "tmpfs" mount. Docker overlay 2 storage driver uses features of the underlying file system, so basically what you need to do is have a huge tmpfs volume mounted as `/var/lib/docker`. And another tmpfs volume (or a folder of it) used to keep checked out code, another one for `/tmp`, another one for "HOME" directory of the user (a lot of caches are stroed there)  and another one (or shared with HOME) for any logs/artifacts that are produced during the build . That's pretty much it :). The volumes just need to be big enough to keep all the images, node_modules, logs and everything that is produced. 64GB should be more than enough. 

It's surprising that it works for Docker - but it does actually, it is supper-reliable and many times faster than using disk. Docker Builds using such tmpfs-backed docker are lightning-fast. The big drawback of it it in normal circumstances is - of-course - that you loose everything on reboot, however this is not a problem for CI machines. They are ephemeral by nature, and actually it's a very nice property to have a clean state for every job you run. 

One of the big problems with CI machines is to keep them "healthy" - i.e.  avoid uncontrollable growth of the disk space/memory used over time. But with ephemeral machines that have maximum lifetime of 24 hrs (which I believe is the case on AWS) this is less of a problem I think.

There are two possibilities you can take here:

1) you wipe (basically delete and recreate) the tmpfs volumes for sources and logs before/after every job and keep the docker volume as long as the machine is up.

This requires periodic docker system prune and if we have ephemeral machines, this might be a viable option as those will be restarted often anyway. This will speed up the builds running on the same machine a little due to docker caching but it also introduces the risk of the memory going uncontrollably up - for example if we have a new python base image released and heavy docker changed in some PR, until reboot all the image 'variants' will be kept in memory and due to the size of our CI images it can easily baloon to 20-30 GB or more very quickly. I usually avoid the setup where there is any caching (even docker image caching) between jobs on a local CI build machine though so I usually prefer option 2). This makes all the "left-overs", "permissions left", "management" and "uncontrollable growth" problems go away without spending a minute on analyzing/solving them.

2) you wipe all the tmpfs volumes before every job: docker root, sources, log/artifact volumes.

This looks counter-intuitive at first (we do not do local caching between the jobs).  But it is usually the best option as it provides a "True Clean State"  (TM) :) for  every job.  Our CI scripts will cope very well with this in fact (I have foreseen this scenario from the beginning). We are using GitHub Registry/Packages as image cache and we do not need any more cache. The scripts use "PULL" cache mechanism to build the images, which means that we always pull the latest "golden" image from cache (latest that passed the tests on master), before the image is built, then we use `--cache-from` directive to use it as starting point. In VAST majority of cases this will be really fast. Pulling the CI image is about 1 minute now, and rebuilding it (using a disk-based docker) is 3 minutes for simple changes, 16 minutes for complete rebuild (when first line of the Docker file changed). This will likely go down to a little less than 1 minute to pull (speed of network will be the only limit and in AWS machine it will be super fast) and likely (I am guessing now) around 30 seconds (sources changed) - 5 minutes (whole Dockerfile changed) to do rebuild in memory-backed docker engine. I think this 1 minute pull is really small comparing to all the benefits you get by "True Clean State". 

Using "tmpfs" in this case for all volumes is really nice because the wipe is immediate and always "clean" - you just drop the volume and recreate it - which takes less than a second usually. Basically the goal is that other than starting the OS and loading libraries and binaries in memory there is 0 activity of the disk. You can monitor that during the build and see what files are accessed and put all those on tmpfs as well if you see any activiy. Your goal should be that everything is in memory.

Then you have a blasing fast (and cheap at that) CI :).

**Comment by ashb** [comment]
I'm running now with tmpfs for `/var/lib/docker`, `/tmp` and the runner `_work` dir. No changes needed on our code side to support it, all just a machine config change.

File: .github/workflows/ci.yml
Comment (potiuk) [inline .github/workflows/ci.yml:122]: Nice

**Review by kaxil — APPROVED** [review]

**Review by potiuk — APPROVED** [review]
Looks good, except the fcntl part which likely can be removed for the in-memory run.

**Comment by ashb** [comment]
(Doesn't look like tmpfs helps much with the "main" test suite)

**Comment by potiuk** [comment]
I will take a look tomorrow

**Comment by ashb** [comment]
Just had the "build provider wheel` step fail with no obvious error code or exit, so it looks like the fcntl is still needed (maybe?)

**Comment by ashb** [comment]
https://bugs.python.org/issue29130 :disappointed:  "The 120 code was added in Issue 5319, as a way of indicating a problem in the final stages of the interpreter exiting

https://pipelines.actions.githubusercontent.com/dG18C1r9Veh0Ng8eypyIHvAXk1Mz6Bvz6mj1myj6BZplhvMhGo/_apis/pipelines/1/runs/54614/signedlogcontent/46?urlExpires=2021-02-08T20%3A59%3A13.7186789Z&urlSigningMethod=HMACV1&urlSignature=ksgS8h%2BHRB6BE7868RzngkEs%2Fn4tek70RT%2BwKVnJjnk%3D

**Comment by ashb** [comment]
✅ ✅ ✅ ✅ ✅ ✅ ✅ ✅!

**Comment by potiuk** [comment]
:tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada:

**Pull request merged**
